content/master/concepts/composition-functions.md

```diff
@@ -1,1026 +1,637 @@
-          name: platform-ref-gcp-db-p9wrj
-          uid: 96623f41-be2e-4eda-84d4-9668b48e284d
-        resourceVersion: "9951"
-        uid: ab5dafbe-2bc8-47ea-8b5b-9bcb40183e45
-      spec:
-        forProvider:
-          instance: platform-ref-gcp-db-p9wrj-tvvtg
-          project: example
-        providerConfigRef:
-          name: default
-    # Any desired connection details for the new db-user composed resource.
-    # Desired connection details can be FromValue, FromFieldPath, or
-    # FromConnectionSecretKey, just like their P&T Composition equivalents.
-    connectionDetails:
-    - name: password
-      type: FromValue
-      value: very-secret
-    # Any desired readiness checks for the new db-user composed resource.
-    # Desired readiness checks can be NonEmpty, MatchString, MatchInteger, or
-    # None, just like their P&T Composition equivalents.    
-    readinessChecks:
-    - type: None
-# An optional array of results.
-results:
-- severity: Normal
-  message: "Successfully composed GCP SQL user"
-```
-{{< /expand >}}
+# Edit the template to add your function's logic
+$ vim fn.go
 
-### An example Function
+# Build the function.
+$ docker build . --quiet --tag runtime
+sha256:2c31b0f7a34b34ba5b0b2dacc94c360d18aca1b99f56ca4f40a1f26535a7c1c4
 
-You can write a Composition Function using any programming language that can be
-containerized, or existing tools like Helm or Kustomize.
+# Package the function.
+$ crossplane xpkg build -f package --embed-runtime-image=runtime
 
-Here's a Python Composition Function that doesn't create any new desired
-resources, but instead annotates any existing desired resources with a quote.
-Because this function accesses the internet it needs to be run with the `Runner`
-network policy.
+# Push the function package to xpkg.upbound.io.
+$ crossplane xpkg push -f package/*.xpkg crossplane-contrib/function-example:v0.1.0
+```
 
-```python
-import sys
+{{<hint "tip">}}
+Crossplane has
+[language specific guides]({{<ref "../../knowledge-base/guides">}}) to writing
+a composition function. Refer to the guide for your preferred language for a
+more detailed guide to writing a function.
+{{</hint>}}
+
+When you're writing a composition function it's useful to know how composition
+functions work. Read the next section to learn
+[how composition functions work](#how-composition-functions-work).
+
+## How composition functions work
+
+Each composition function is actually a [gRPC](https://grpc.io) server. gRPC is
+a high performance, open source remote procedure call (RPC) framework. When you
+[install a function](#install-a-composition-function) Crossplane deploys the
+function as a gRPC server. Crossplane encrypts and authenticates all gRPC
+communication.
+
+You don't have to be a gRPC expert to write a function. Crossplane's function
+SDKs setup gRPC for you. It's useful to understand how Crossplane calls your
+function though, and how your function should respond.
+
+```mermaid
```

Comment:
This looks great to me in light mode, but not so much in dark mode. It renders the line labels as dark grey on a black background.

Reply:
> this image is hard to read when using dark mode

@JonasKs Yeah, this is a known issue with mermaid graph rendering. We added it in this PR but haven't done any styling yet. I'd like to fix that a subsequent PR as it will probably need @plumbis expertise to do it right.

Reply:
Nevermind, looks like @plumbis added a commit that fixes this. 🎉